### PR TITLE
[Order] Order __constructor dependencies by type alphabetically

### LINKED ISSUE #3714: [CodingStyle] Order __constructor dependencies by type A-Z
```diff
public function __construct(
+       LatteAndTwigFinder $latteAndTwigFinder,
        LatteToTwigConverter $latteToTwigConverter,
-       SymfonyStyle $symfonyStyle,
-       LatteAndTwigFinder $latteAndTwigFinder,
        SmartFileSystem $smartFileSystem,
+        SymfonyStyle $symfonyStyle, 
   ) {
```

### PR DESCRIPTION
**Pull request by dobryy**
Closes #3714

### CODE REVIEW

File: rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+        foreach ($node->params as $param) {
+            if ($this->isDefaultValueParam($param)) {
+                return true;
+            }
+
+            if ($this->isPrimitiveDataTypeParam($param)) {
+                return true;
+            }
+        }
```
Comment (dobryy) [inline rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:131]: @TomasVotruba Originally, I made two methods `hasDefaultValueParam` and `hasPrimitiveDataTypeParam` and they contained their own foreach loop but then I decided to have one loop. Please let me know which approach is better.

Comment (TomasVotruba) [reply rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php]: I think both are fine :+1:

**Comment by dobryy** [comment]
@TomasVotruba phpstan says that in a couple of years (when you'll be around 70 y.o. :D) you won't understand whats going on in Rector.

```
------------------------------------------------------------------------
 src/Application/RectorApplication.php:116
 ------------------------------------------------------------------------
 - '#Cognitive complexity for "Rector\\Core\\Application\\RectorApplication\:\:runOnFileInfos\(\)" is 12, keep it under 9#'
 ------------------------------------------------------------------------
```

This issue shouldn't be caused by my changes.

File: rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        return $node;
+    }
+
+    /**
+     * @return Param[]
+     */
+    private function getSortedParams(ClassMethod $classMethod): array
+    {
+        $params = $classMethod->getParams();
+        usort($params, function (Param $a, Param $b) {
```
Comment (TomasVotruba) [inline rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:84]:
```suggestion
        usort($params, function (Param $firstParam, Param $secondParam) {
```

Names should respect types as normal params do.

Possible place for next naming rule ;)

File: rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    private function isDefaultValueParam(Param $param): bool
+    {
+        return $param->default instanceof Expr;
+    }
+
+    private function shouldSkip(ClassMethod $classMethod): bool
+    {
+        if ($this->getName($classMethod) !== '__construct') {
```
Comment (TomasVotruba) [inline rules/order/src/Rector/ClassMethod/OrderConstructorDependenciesByTypeAlphabeticallyRector.php:107]:
```suggestion
        if (! $this->isName($classMethod, '__construct')) {
```

For these case it's better to use *check* method

**Comment by dobryy** [comment]
> Looks good. How does it run on Rector code?

Forgot the most interesting part :D it was too late for me.
Had to fix one issue to make it work.

Here is the commit where I applied the rule https://github.com/rectorphp/rector/pull/3786/commits/b023dbe7acff2c6a8c7de47aeb8521cabfdee000

Here is the list of classes in rector which have nullable type or expr for param (mostly they have the last param with default value of empty array):
```
Rector\Core\Console\Application
Rector\Core\Exception\ShouldNotHappenException
Rector\Core\Exclusion\ExclusionManager
Rector\Core\PhpParser\Builder\UseBuilder
Rector\Core\PhpParser\NodeTraverser\RectorNodeTraverser
Rector\Core\PhpParser\Printer\BetterStandardPrinter
Rector\Core\Rector\Architecture\Factory\NewObjectToFactoryCreateRector
Rector\Core\Rector\Argument\ArgumentAdderRector
Rector\Core\Rector\Argument\ArgumentDefaultValueReplacerRector
Rector\Core\Rector\Argument\ArgumentRemoverRector
Rector\Core\Rector\Argument\SwapFuncCallArgumentsRector
Rector\Core\Rector\Assign\PropertyAssignToMethodCallRector
Rector\Core\Rector\ClassLike\RemoveTraitRector
Rector\Core\Rector\ClassMethod\AddMethodParentCallRector
Rector\Core\Rector\ClassMethod\AddReturnTypeDeclarationRector
Rector\Core\Rector\ClassMethod\ChangeContractMethodSingleToManyRector
Rector\Core\Rector\ClassMethod\WrapReturnRector
Rector\Core\Rector\Class_\AddInterfaceByTraitRector
Rector\Core\Rector\Class_\ParentClassToTraitsRector
Rector\Core\Rector\Constant\RenameClassConstantsUseToStringsRector
Rector\Core\Rector\FuncCall\FunctionToNewRector
Rector\Core\Rector\FuncCall\RemoveFuncCallArgRector
Rector\Core\Rector\FuncCall\RemoveIniGetSetFuncCallRector
Rector\Core\Rector\Function_\FunctionToMethodCallRector
Rector\Core\Rector\Function_\FunctionToStaticCallRector
Rector\Core\Rector\Interface_\MergeInterfacesRector
Rector\Core\Rector\Interface_\RemoveInterfacesRector
Rector\Core\Rector\MethodBody\NormalToFluentRector
Rector\Core\Rector\MethodCall\MethodCallToAnotherMethodCallWithArgumentsRector
Rector\Core\Rector\MethodCall\MethodCallToReturnRector
Rector\Core\Rector\MethodCall\ServiceGetterToConstructorInjectionRector
Rector\Core\Rector\Namespace_\PseudoNamespaceToNamespaceRector
Rector\Core\Rector\New_\NewToStaticCallRector
Rector\Core\Rector\Property\InjectAnnotationClassRector
Rector\Core\Rector\Property\PropertyToMethodRector
Rector\Core\Rector\Property\RenamePropertyRector
Rector\Core\Rector\StaticCall\StaticCallToFunctionRector
Rector\Core\Rector\StaticCall\SwapClassMethodArgumentsRector
Rector\Core\Rector\String_\StringToClassConstantRector
Rector\Core\Rector\Visibility\ChangeConstantVisibilityRector
Rector\Core\Rector\Visibility\ChangeMethodVisibilityRector
Rector\Core\Rector\Visibility\ChangePropertyVisibilityRector
Rector\Core\ValueObject\AssignAndRootExpr
Rector\Autodiscovery\Rector\FileSystem\MoveServicesBySuffixToDirectoryRector
Rector\Autodiscovery\Rector\FileSystem\MoveValueObjectsToValueObjectDirectoryRector
Rector\CakePHP\Rector\MethodCall\ArrayToFluentCallRector
Rector\CakePHP\Rector\MethodCall\ModalToGetSetRector
Rector\CakePHP\Rector\MethodCall\RenameMethodCallBasedOnParameterRector
Rector\CodingStyle\Rector\ClassMethod\ReturnArrayClassMethodToYieldRector
Rector\CodingStyle\Rector\ClassMethod\YieldClassMethodToArrayClassMethodRector
Rector\CodingStyle\Rector\FuncCall\ConsistentPregDelimiterRector
Rector\CodingStyle\Rector\FuncCall\FunctionCallToConstantRector
Rector\CodingStyle\Rector\MethodCall\PreferThisOrSelfMethodCallRector
Rector\Decouple\Rector\DecoupleClassMethodToOwnClassRector
Rector\Doctrine\Rector\Class_\AddEntityIdByConditionRector
Rector\Doctrine\Rector\MethodCall\EntityAliasToClassConstantReferenceRector
Rector\Legacy\Rector\Include_\AddTopIncludeRector
Rector\MagicDisclosure\Rector\Assign\GetAndSetToMethodCallRector
Rector\MagicDisclosure\Rector\ClassMethod\ReturnThisRemoveRector
Rector\MagicDisclosure\Rector\Isset_\UnsetAndIssetToMethodCallRector
Rector\MagicDisclosure\Rector\MethodCall\DefluentMethodCallRector
Rector\MagicDisclosure\Rector\String_\ToStringToMethodCallRector
Rector\Order\Rector\Class_\OrderPublicInterfaceMethodRector
Rector\Php55\Rector\String_\StringClassNameToClassConstantRector
Rector\Php70\EregToPcreTransformer
Rector\Php71\Rector\Name\ReservedObjectRector
Rector\Php74\Rector\Function_\ReservedFnFunctionRector
Rector\Php74\Rector\Property\TypedPropertyRector
Rector\PHPUnit\Rector\Class_\ArrayArgumentInTestToDataProviderRector
Rector\RemovingStatic\Rector\Class_\NewUniqueObjectToEntityFactoryRector
Rector\RemovingStatic\Rector\Class_\PHPUnitStaticToKernelTestCaseGetRector
Rector\RemovingStatic\Rector\Class_\PassFactoryToUniqueObjectRector
Rector\RemovingStatic\Rector\Class_\StaticTypeToSetterInjectionRector
Rector\Renaming\Rector\Annotation\RenameAnnotationRector
Rector\Renaming\Rector\Class_\RenameClassRector
Rector\Renaming\Rector\ConstFetch\RenameConstantRector
Rector\Renaming\Rector\Constant\RenameClassConstantRector
Rector\Renaming\Rector\FuncCall\RenameFuncCallToStaticCallRector
Rector\Renaming\Rector\Function_\RenameFunctionRector
Rector\Renaming\Rector\MethodCall\RenameMethodRector
Rector\Renaming\Rector\MethodCall\RenameStaticMethodRector
Rector\Renaming\Rector\Namespace_\RenameNamespaceRector
Rector\Restoration\Rector\Class_\RemoveUselessJustForSakeInterfaceRector
Rector\Restoration\Rector\Namespace_\CompleteImportForPartialAnnotationRector
Rector\Restoration\Rector\New_\CompleteMissingDependencyInNewRector
Rector\SOLID\Rector\Class_\MultiParentingToAbstractDependencyRector
Rector\Symfony\Rector\Class_\ChangeFileLoaderInExtensionAndKernelRector
Rector\Symfony\Rector\FrameworkBundle\ContainerGetToConstructorInjectionRector
Rector\Symfony\Rector\FrameworkBundle\GetToConstructorInjectionRector
Rector\Twig\Rector\SimpleFunctionAndFilterRector
Rector\TypeDeclaration\Rector\ClassMethod\AddParamTypeDeclarationRector
Rector\TypeDeclaration\Rector\FunctionLike\ReturnTypeDeclarationRector
Rector\PHPStanExtensions\Rule\RequireStringArgumentInMethodCallRule
Rector\BetterPhpDocParser\PhpDocParser\BetterPhpDocParser
Rector\BetterPhpDocParser\ValueObject\TagValueNodeConfiguration
Rector\FileSystemRector\FileSystemFileProcessor
Rector\FileSystemRector\Rector\Removing\RemoveProjectFileRector
Rector\NodeNameResolver\NodeNameResolver
Rector\PHPUnit\ValueObject\ParamAndArgValueObject

```

**Comment by dobryy** [comment]
Is it okay that the progress stops on 79% with success message?

```
  8384/10490 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░]  79%

                                                                                                                        
 [OK] Rector is done!   
```

**Comment by TomasVotruba** [comment]
> @TomasVotruba phpstan says that in a couple of years (when you'll be around 70 y.o. :D) you won't understand whats going on in Rector.

:laughing:  

Fixed on master. Rebase should help :)

**Comment by TomasVotruba** [comment]
> Here is the list of classes in rector which have nullable type or expr for param (mostly they have the last param with default value of empty array)

If you try to re-order them too, what will break? 
Standalone commit should show us in the CI.


I just realized classes from *ValueObject* should be skipped, there are created manually.

**Comment by TomasVotruba** [comment]
Great job, thanks :+1: 

I love the order :sunglasses:

**Comment by TomasVotruba** [comment]
> Is it okay that the progress stops on 79% with success message?
> 
> ```
>   8384/10490 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓░░░░░░]  79%
> 
>                                                                                                                         
>  [OK] Rector is done!   
> ```

I just noticed this comment, got lost in the crowd.
I'll fix it :) thanks for reporting!

**Pull request merged**